aws/rust-runtime/aws-runtime/src/auth/sigv4a.rs

```diff
@@ -0,0 +1,318 @@
+        `expires_in` duration because the credentials used to sign it will expire first.";
+
+/// Auth scheme ID for SigV4a.
+pub const SCHEME_ID: AuthSchemeId = AuthSchemeId::new("sigv4a");
+
+/// SigV4a auth scheme.
+#[derive(Debug, Default)]
+pub struct SigV4aAuthScheme {
+    signer: SigV4aSigner,
+}
+
+impl SigV4aAuthScheme {
+    /// Creates a new `SigV4aHttpAuthScheme`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl AuthScheme for SigV4aAuthScheme {
+    fn scheme_id(&self) -> AuthSchemeId {
+        SCHEME_ID
+    }
+
+    fn identity_resolver(
+        &self,
+        identity_resolvers: &dyn GetIdentityResolver,
+    ) -> Option<SharedIdentityResolver> {
+        identity_resolvers.identity_resolver(self.scheme_id())
+    }
+
+    fn signer(&self) -> &dyn Signer {
+        &self.signer
+    }
+}
+
+/// SigV4a HTTP request signer.
+#[derive(Debug, Default)]
+pub struct SigV4aSigner;
+
+impl SigV4aSigner {
+    /// Creates a new signer instance.
+    pub fn new() -> Self {
+        Self
+    }
+
+    fn settings(operation_config: &SigV4OperationSigningConfig) -> SigningSettings {
+        super::settings(operation_config)
+    }
+
+    fn signing_params<'a>(
+        settings: SigningSettings,
+        identity: &'a Identity,
+        operation_config: &'a SigV4OperationSigningConfig,
+        request_timestamp: SystemTime,
+    ) -> Result<v4a::SigningParams<'a, SigningSettings>, SigV4SigningError> {
+        if let Some(expires_in) = settings.expires_in {
+            if let Some(&identity_expiration) = identity.expiration() {
+                let presigned_expires_time = request_timestamp + expires_in;
+                if presigned_expires_time > identity_expiration {
+                    tracing::warn!(EXPIRATION_WARNING);
+                }
+            }
+        }
+
+        Ok(v4a::SigningParams::builder()
+            .identity(identity)
+            .region_set(
+                operation_config
+                    .region_set
+                    .as_ref()
+                    .ok_or(SigV4SigningError::MissingSigningRegionSet)?
+                    .as_ref(),
+            )
+            .name(
+                operation_config
+                    .name
+                    .as_ref()
+                    .ok_or(SigV4SigningError::MissingSigningName)?
+                    .as_ref(),
+            )
+            .time(request_timestamp)
+            .settings(settings)
+            .build()
+            .expect("all required fields set"))
+    }
+
+    fn extract_operation_config<'a>(
+        auth_scheme_endpoint_config: AuthSchemeEndpointConfig<'a>,
+        config_bag: &'a ConfigBag,
+    ) -> Result<Cow<'a, SigV4OperationSigningConfig>, SigV4SigningError> {
+        let operation_config = config_bag
+            .load::<SigV4OperationSigningConfig>()
+            .ok_or(SigV4SigningError::MissingOperationSigningConfig)?;
```

Comment:
I can see it being confusing to people to receive a Sigv4 error when doing a Sigv4a thing. Since implementing this feature, I really see them as the same thing, except that:
- There's potentially more than one region
- There's a requirement for a signing key

Everything else is the same. I think that if I was in the business of renaming things, I'd refer to an `AWS Signer` in the code and then name 'sub-signers' as different versions of the AWS signer.